packages/client/src/runtime/core/engines/client/TransactionManager.ts

```diff
@@ -0,0 +1,169 @@
+import Debug from '@prisma/debug'
+import { DriverAdapter, Transaction } from '@prisma/driver-adapter-utils'
+import { assertNever } from '@prisma/internals'
+import crypto from 'crypto'
+
+import type * as Tx from '../common/types/Transaction'
+import {
+  TransactionAlreadyCommittedError,
+  TransactionClosedError,
+  TransactionDriverAdapterError,
+  TransactionInternalConsistencyError,
+  TransactionManagerError,
+  TransactionNotFoundError,
+  TransactionRolledBackError,
+  TransactionTimedOutError,
+} from './TransactionManagerErrors'
+
+const MAX_CLOSED_TRANSACTIONS = 100
+
+type TransactionWrapper = {
+  id: string
+  status: 'waiting' | 'running' | 'committed' | 'rolled_back' | 'timed_out'
+  timeout?: NodeJS.Timeout
+  transaction?: Transaction
+}
+
+const debug = Debug('prisma:client:transactionManager')
+
+export class TransactionManager {
+  // The map of active transactions.
+  private transactions: Map<string, TransactionWrapper> = new Map()
+  // List of last closed transactions. Max MAX_CLOSED_TRANSACTIONS entries.
+  // Used to provide better error messages than a generic "transaction not found".
+  private closedTransactions: TransactionWrapper[] = []
+  private driverAdapter: DriverAdapter
+
+  constructor(driverAdapter: DriverAdapter) {
+    this.driverAdapter = driverAdapter
+  }
+
+  async startTransaction(options: Tx.Options): Promise<Tx.InteractiveTransactionInfo<undefined>> {
+    // Supplying timeout default values is cared for upstream already.
+    if (!options.timeout) throw new TransactionManagerError('Timeout is required')
+    if (!options.maxWait) throw new TransactionManagerError('maxWait is required')
+    // TODO: care about transaction isolation levels?
```

Comment:
❓ I am not sure how we actually plan to handle transaction isolation levels with compiler approach. Somehow we'll have to pass this info to the query compiler? Or shall this transaction manager or the driver adapter inject the necessary SQL statements (and know about dialect/db differences?...)? 🤔